### PR TITLE
Diarization error rate and identification error rate

### PR DESCRIPTION
**Pull request by audeerington**
This adds `audmetric.diarization_error_rate()` and `audmetric.identification_error_rate()`.

image

image

The implementation is based on the metrics `pyannote.metrics.diarization.GreedyDiarizationErrorRate` and `pyannote.metrics.identification.IdentificationErrorRate` from [pyannote.metrics](https://pyannote.github.io/pyannote-metrics/index.html). We also use these metrics to generate some test cases.

## Summary by Sourcery

Implement speaker diarization and identification error rate metrics based on pyannote.metrics, introduce supporting utilities for segment mapping, and add comprehensive testing with reference data.

New Features:
- Add diarization_error_rate function with greedy mapping
- Add identification_error_rate function computing confusion, false alarms, and misses over segmented audio

Enhancements:
- Introduce internal helper functions for segment boundary detection, cooccurrence mapping, and segment intersection

Build:
- Update pytest configuration to ignore new test asset directories

Documentation:
- Add Bredin2017 citation for metric definitions

Tests:
- Add unit tests for DER and IER covering edge cases and pyannote-metrics reference comparisons

### CODE REVIEW

**Comment by audeerington** [comment]
@felixbur I added a fix to the implementation, and tried out your example again and it now works for me.

Specifically, I added a step to the beginning of the diarization error rate function, similar to how it is done my `pyannote.metrics`, that gives the truth and prediction new, unique labels before applying a mapping. I thought this would not be necessary when first implementing it, but I now saw that it is required for the tests to pass.

File: tests/assets/error_rates/ier/4/truth.csv [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+f1.wav,0 days 00:09:56.042999250,0 days 00:09:57.663546391,label_8
+f1.wav,0 days 00:09:58.503467146,0 days 00:09:59.446044435,label_8
+f1.wav,0 days 00:09:59.490726946,0 days 00:09:59.994680496,label_8
+f1.wav,0 days 00:03:27.246443295,0 days 00:03:30.783815516,label_9
+f1.wav,0 days 00:04:56.877475153,0 days 00:04:59.480081360,label_9
+f1.wav,0 days 00:06:33.916057262,0 days 00:06:37.236907472,label_9
+f1.wav,0 days 00:09:34.694550346,0 days 00:09:37.209150846,label_9
+f1.wav,0 days 00:09:54.695372836,0 days 00:09:57.251718489,label_9
+f1.wav,0 days 00:09:58.727106762,0 days 00:09:59.329005760,label_9
+f1.wav,0 days 00:09:59.403043104,0 days 00:09:59.991207172,label_9
```
Comment (hagenw) [inline tests/assets/error_rates/ier/4/truth.csv]: All test assets have a segmented index with a single file. Do we have a test case that ensures it works if you have more than one file?

File: tests/test_error_rates.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+            0.42 / 2.95,
+        ),
+        # Confusions, misses and false alarms
+        (
+            pd.Series(
+                index=audformat.segmented_index(
+                    files=["f1.wav"] * 2 + ["f2.wav"] * 4 + ["f3.wav"],
+                    starts=[0, 0.1, 0.2, 0.4, 1.0, 3.5, 0.0],
+                    ends=[0.1, 0.2, 0.35, 0.8, 3.0, 4.0, 1.0],
+                ),
```
Comment (hagenw) [inline tests/test_error_rates.py]: OK, here are cases with different files. So I guess this is sufficient?

File: tests/assets/error_rates/ier/4/truth.csv (hunk 1 above)
Comment (audeerington) [reply tests/assets/error_rates/ier/4/truth.csv]: Yes, the reason for that is that the pyannote error rate metrics are applied on each file individually, so you only pass information of start, end, and label of each segment. As you already pointed out there are other tests that cover multiple files.

Comment (hagenw) [reply tests/assets/error_rates/ier/4/truth.csv]: OK, thanks.

File: audmetric/core/api.py [hunk 3]
```diff
@@ -290,6 +290,89 @@ def detection_error_tradeoff(
...
+    :math:`\text{false alarm}` is the total duration of predictions
+    without an overlapping ground truth,
+    :math:`\text{miss}` is the total duration of ground truth
+    without an overlapping prediction,
+    and :math:`\text{total}` is the total duration of ground truth segments.
+
+    This metric is computed the same way
+    as :func:`audmetric.identification_error_rate`,
+    but first creates a one-to-one mapping between
+    truth and prediction labels.
```
[the remark this answers is not in the document]
Comment (hagenw) [reply audmetric/core/api.py]: OK, thanks for the clarification.

Even for cases in which IER and DER would return the same results, IER should also be faster as there is no need to match the labels.

But I would settle here along the lines you proposed and just highlight more in the docstring of IER and DER when those functions should be used.

File: audmetric/core/api.py [hunk 4]
```diff
@@ -1246,6 +1329,134 @@ def fscore_per_class(
...
+    total_duration = 0
+    total_confusion = 0
+    total_misses = 0
+    total_false_alarm = 0
+    files = (
+        truth.index.get_level_values(FILE)
+        .unique()
+        .union(prediction.index.get_level_values(FILE).unique())
+    )
+    for file in files:
```
[the remark this answers is not in the document]
Comment (hagenw) [reply audmetric/core/api.py]: Great, I think it is fine to stay without the verbose option for now and only add it if users complain.

File: audmetric/core/api.py [hunk 5]
```diff
@@ -2332,26 +2355,58 @@ def _diarization_mapper(
...
+        files = truth.index.get_level_values(FILE).unique()
+        file_mappings = audeer.run_tasks(
+            _diarization_mapper,
+            params=[
+                (
+                    (
+                        truth[truth.index.get_level_values(FILE) == file],
+                        prediction[prediction.index.get_level_values(FILE) == file],
+                    ),
+                    {"individual_file_mapping": False},
```
[the remark this answers is not in the document]
Comment (hagenw) [reply audmetric/core/api.py]: Sorry, I overlooked that it is calling itself again.

**Comment by hagenw** [comment]
> > What value is returned when we do not have overlap between the files?
> 
> If there is no overlap, then all segments of the prediction are false alarms and all segments of the truth are misses. That means the error rate is `(duration_false_alarm+duration_misses)/duration_truth` (`duration_misses` is the same as `duration_truth` in this case).

OK, `duration_false_alarm` is given by the length of prediction, which can be very short, but then the value would always be >=1.0, so we should be fine here.

File: audmetric/core/api.py [hunk 6]
```diff
@@ -376,6 +376,8 @@ def diarization_error_rate(
             "The truth and prediction "
             "should be a pandas Series with a segmented index conform to audformat."
         )
+    # Warn user if there are no common files
+    _check_common_files(truth, prediction)
```
Comment (hagenw) [inline audmetric/core/api.py]: It's not ideal that we need to check twice for this when running calculating DER, one time here and one time inside `identification_error_rate()`

File: audmetric/core/api.py [hunk 7]
```diff
@@ -1446,6 +1448,9 @@ def identification_error_rate(
...
```
[the remark this answers is not in the document]
Comment (audeerington) [reply audmetric/core/api.py]: I applied the suggestion and also removed the extra check in the DER function.

**Comment by hagenw** [comment]
Feel free to merge here.
